cmd/juju/commands/bundle.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package commands
+
+import (
+	"fmt"
+
+	"github.com/juju/bundlechanges"
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"gopkg.in/juju/charm.v6-unstable"
+
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/environs/config"
+)
+
+// deployBundle deploys the given bundle data using the given API client and
+// charm store client. The deployment is not transactional, and its progress is
+// notified using the given command context.
+func deployBundle(data *charm.BundleData, client *api.Client, csclient *csClient, repoPath string, conf *config.Config, ctx *cmd.Context) error {
```

Comment:
I wonder whether it might be better to provide notifications through an interface rather than directly to a cmd.Context - that way this function is potentially usable server-side too.